crates/provider-keyvalue-redis/src/lib.rs

```diff
@@ -381,6 +553,153 @@ impl Provider for KvRedisProvider {
+        &self,
+        LinkConfig {
+            target_id,
+            config,
+            secrets,
+            link_name,
+            wit_metadata: (_, _, interfaces),
+            ..
+        }: LinkConfig<'_>,
+    ) -> anyhow::Result<()> {
+        let url = secrets
+            .keys()
+            .find(|k| k.eq_ignore_ascii_case(CONFIG_REDIS_URL_KEY))
+            .and_then(|url_key| config.get(url_key))
+            .or_else(|| {
+                warn!("Redis connection URLs can be sensitive. Consider using secrets to pass this value.");
+                config.keys()
+                    .find(|k| k.eq_ignore_ascii_case(CONFIG_REDIS_URL_KEY))
+                    .and_then(|url_key| config.get(url_key))
+            })
+            .map_or(DEFAULT_CONNECT_URL, |v| v);
+
+        let client = match redis::Client::open(url.to_string()) {
+            Ok(client) => {
+                info!(url, "Established link at receive_link_config_as_source");
+                client
+            }
+            Err(err) => {
+                warn!(?err, "Failed to create Redis client for target [{target_id}]. Key-value operations will fail.");
+                bail!("Failed to create Redis client");
+            }
+        };
+
+        let mut conn = client.get_connection_manager().await.map_err(|e| {
+            error!("Failed to get async connection: {e}");
+            anyhow::anyhow!("Failed to get async connection: {}", e)
+        })?;
+
+        let component_id: Arc<str> = target_id.into();
+        let wrpc = get_connection()
+            .get_wrpc_client(&component_id)
+            .await
+            .context("failed to construct wRPC client")?;
+        if interfaces.contains(&"watcher".to_string()) {
+            let wrpc = Arc::new(wrpc);
+            let wrpc_for_task = wrpc.clone();
+            // Set keyspace notifications
+            redis::cmd("CONFIG")
+                .arg("SET")
+                .arg("notify-keyspace-events")
+                .arg("KEA")
+                .query_async::<_, ()>(&mut conn)
+                .await
+                .map_err(|e| {
+                    error!("Failed to set keyspace notifications: {e}");
+                    anyhow::anyhow!("Failed to set keyspace notifications: {}", e)
+                })?;
+
+            // Start watching for keyspace events
+            let client_clone = client.clone();
+            let self_clone = self.clone();
+            let mut conn_clone = conn.clone();
+            let task = tokio::spawn(async move {
+                let mut pubsub = match client_clone.get_async_pubsub().await {
+                    Ok(pubsub) => pubsub,
+                    Err(e) => {
+                        error!("Failed to get pubsub connection: {}", e);
+                        return;
+                    }
+                };
+                pubsub.psubscribe("__keyevent@0__:set").await.unwrap();
+                pubsub.psubscribe("__keyevent@0__:del").await.unwrap();
```

Comment:
I have a feeling some of this might get replaced with the `WATCH` command, but just a note to change `unwrap` calls where possible to `.context("error info")` after importing the `anyhow::Context` crate, so that we don't panic 😄